relay-general/src/store/normalize.rs

```diff
@@ -291,166 +158,352 @@ impl<'a> NormalizeProcessor<'a> {
+            );
+        }
+    }
+}
+
+/// Backfills IP addresses in various places.
+fn normalize_ip_addresses(event: &mut Event, client_ip: Option<&IpAddr>) {
+    // NOTE: This is highly order dependent, in the sense that both the statements within this
+    // function need to be executed in a certain order, and that other normalization code
+    // (geoip lookup) needs to run after this.
+    //
+    // After a series of regressions over the old Python spaghetti code we decided to put it
+    // back into one function. If a desire to split this code up overcomes you, put this in a
+    // new processor and make sure all of it runs before the rest of normalization.
+
+    // Resolve {{auto}}
+    if let Some(client_ip) = client_ip {
+        if let Some(ref mut request) = event.request.value_mut() {
+            if let Some(ref mut env) = request.env.value_mut() {
+                if let Some(&mut Value::String(ref mut http_ip)) = env
+                    .get_mut("REMOTE_ADDR")
+                    .and_then(|annotated| annotated.value_mut().as_mut())
+                {
+                    if http_ip == "{{auto}}" {
+                        *http_ip = client_ip.to_string();
                     }
                 }
             }
         }
 
-        Ok(())
+        if let Some(ref mut user) = event.user.value_mut() {
+            if let Some(ref mut user_ip) = user.ip_address.value_mut() {
+                if user_ip.is_auto() {
+                    *user_ip = client_ip.to_owned();
+                }
+            }
+        }
     }
 
-    fn normalize_user_agent(&self, _event: &mut Event) {
-        if self.config.normalize_user_agent.unwrap_or(false) {
-            #[cfg(feature = "uaparser")]
-            user_agent::normalize_user_agent(_event);
-
-            #[cfg(not(feature = "uaparser"))]
-            panic!("relay not built with uaparser feature");
+    // Copy IPs from request interface to user, and resolve platform-specific backfilling
+    let http_ip = event
+        .request
+        .value()
+        .and_then(|request| request.env.value())
+        .and_then(|env| env.get("REMOTE_ADDR"))
+        .and_then(Annotated::<Value>::as_str)
+        .and_then(|ip| IpAddr::parse(ip).ok());
+
+    if let Some(http_ip) = http_ip {
+        let user = event.user.value_mut().get_or_insert_with(User::default);
+        user.ip_address.value_mut().get_or_insert(http_ip);
+    } else if let Some(client_ip) = client_ip {
+        let user = event.user.value_mut().get_or_insert_with(User::default);
+        // auto is already handled above
+        if user.ip_address.value().is_none() {
+            let platform = event.platform.as_str();
+
+            // In an ideal world all SDKs would set {{auto}} explicitly.
+            if let Some("javascript") | Some("cocoa") | Some("objc") = platform {
+                user.ip_address = Annotated::new(client_ip.to_owned());
+            }
         }
     }
 }
 
+pub fn light_normalize_event(
```

Comment:
nit: We should probably come up with a better name for this function, but I don't have one yet. Maybe `pre_normalize_event`?

Reply:
I hard agree "light normalization" isn't descriptive enough. I don't like `pre_normalize_*` because that indicates the operation happens _before_ normalization, and this function actually performs some normalization. I think that _some_ is what causes troubles for the naming. I'd also like to avoid referring to any filtering in the name, if possible -- the function may be used in another context in the future and shouldn't be tied up to current behavior.

What about something like `minimal_normalize_event`? That indicates there's some normalization happening, it doesn't perform the whole normalization, and the explanation of what it does and the motivation behind it could be added to the docstrings. What do you think?